### PR TITLE
Rework Linux/Kerberos native interop layer

### LINKED ISSUE #25205: System.Data.SqlClient packaging
From https://github.com/dotnet/corefx/pull/25134#discussion_r150415393:

> And for better or worse, we already have some such dependencies, e.g. System.Data.SqlClient isn't part of the .NET Runtime download but it does P/Invoke into the System.Net.Security.Native.so/dylib shim.

We either need to fix that, or we need to add a test coverage to ensure that you can mix and match different versions of .NET Core Runtime (that bundles `System.Net.Security.Native.so/dylib`) with different versions of `System.Data.SqlClient` and document/enforce the supported mixing and matching.

The packaging for `System.Data.SqlClient` is suspect in general: All build configurations in https://github.com/dotnet/corefx/blob/master/src/System.Data.SqlClient/src/Configurations.props are netstandard. How am I going to get the .NET Core shims that the implementation depends on when I install it on Mono?

### LINKED ISSUE #37183: Out-of-band System.Data.SqlClient depends on inbox native shims
SqlClient out-of-band netstandard packages depend on .NET Core native shims:

https://github.com/dotnet/corefx/blob/master/src/System.Data.SqlClient/src/System.Data.SqlClient.csproj#L481-L483

The native shims are not compatible between .NET Core versions, e.g. ImportPrincipalName got changed to ImportTargetName in dotnet/corefx#35383.

This leads to failures when the netstandard SqlClient package is used on a different version of .NET Core than what it was build on, e.g. https://github.com/dotnet/corert/issues/7353

### PR DESCRIPTION
**Pull request by davidsh**
The latest changes to the System.Net.Security.Native shim layer fixed a lot of important
bugs for Linux Kerberos usage. But this created a new problem since SqlClient ships
in out-of-band NuGet packages separate from the .NET Core runtime. SqlClient builds
out of the CoreFx repo and uses the common source includes for Kerberos authentication.
This created an unexpected dependency on the System.Net.Security.Native shim layer.
The recent changes to these API signatures caused problems with different combinations
of SqlClient NuGet packages and .NET Core 2.x versus .NET Core 3.0.

After discussion with the SqlClient team, we decided to rework the changes to these
native APIs so that they would remain compatible across all .NET Core versions.

Long-term, the plan is to implement #36896 to expose a Kerberos API in .NET Core which
could be used by SqlClient and other consumers.

Closes #37183
Closes #25205

### CODE REVIEW

**Comment by jkotas** [comment]
How are we going to test that all possible combinations actually work?

I am wondering whether it would it be better to keep the verbatim 2.1 implementation of the shim APIs, without any changes and big comment at the top of the file that this is for SqlClient compat only and can be deleted once this compat hack does not need to be maintained anymore. And then have the live version of the shim APIs that SqlClient would not have dependency on anymore and that we can be evolved without any constrains.

**Comment by davidsh** [comment]
>How are we going to test that all possible combinations actually work?

I've been manually testing the combinations to make sure things work ok.

>I am wondering whether it would it be better to keep the verbatim 2.1 implementation of the shim APIs, 

The original .NET Core 2.1 implementation of System.Net.Security.Native has bugs that we had to fix in .NET Core 3.0 master such as Negotiate Kerberos to NTLM fallback and channel binding token (CBT).
We plan to port some of these functional fixes in master to the release/2.1 branch in order to unblock some customers that are depending on the .NET Core 2.1 release (which is LTS).  But these fixes and this .NET Core 3.0 master branch PR creates a compatible layer since we will keep the API signatures the same as .NET Core 2.1 for the APIs that SqlClient is using from this layer.

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs [hunk 1]
```diff
@@ -22,7 +22,7 @@ internal partial class SNIProxy
     {
         private const int DefaultSqlServerPort = 1433;
         private const int DefaultSqlServerDacPort = 1434;
-        private const string SqlServerSpnHeader = "MSSQLSvc";
+        private const string SqlServerSpnHeader = "MSSQLSvc/";
```
Comment (wfurt) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs]: should we use '@' here to avoid need to copy & replace later?

Comment (davidsh) [reply src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs]: The "@" format is only for Linux.  The Windows format always use "/". The current code (before this PR) has different files for Windows vs. Linux. This PR undoes that to go back to what it was before I made those change a month ago.

The problem is that this managed code here in CoreFx for SqlClient builds as separate package.  And that package can run against an older .NET Core 2.1 runtime.  And that runtime which uses System.Net.Security.Native doesn't understand the "@" format.  So, right now the latest pre-release packages of System.Data.SqlClient (based on this master branch, .NET Core 3.0) is broken. So, this PR is trying to keep the managed layer the same as what it was in .NET Core 2.1.  

This is why the logic for transformation of "/" to "@" moves to the C layer which doesn't build as a separate (out-of-band) package.

**Pull request merged**